### PR TITLE
Drexel Refactoring of CalcCN

### PR DESCRIPTION
**Pull request by drexeleds**
This Pull request is focused on rewriting the CalcCN and associated (reliant) functions to make them more easily understandably and to reduce computation time. To accomplish this the original code in the GWLF-e python file was extracted to a number of independent functions (n~180). These functions were then vectorized and memoized. The finalized function returns the same results for both the original test4.gms file and for 15 additional gms files. This work was completed to support the wikiwatershed project with funding from the William Penn Foundation and the Stroud Water Research Center.

### CODE REVIEW

**Comment by rajadain** [comment]
Hello, I'm just trying to test out this module locally. Upon attempting installation, I see this message:

```
Collecting gwlf-e from git+git://github.com/drexeleds/gwlf-e@develop#egg=gwlf-e (from -r src/mmw/requirements/base.txt (line 18))
  Cloning git://github.com/drexeleds/gwlf-e (to develop) to /tmp/pip-build-WcRjNG/gwlf-e
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-WcRjNG/gwlf-e/setup.py", line 8, in <module>
        from gwlfe.Input.WaterBudget.AMC5_yesterday_inner import cc as amc5ycc
      File "gwlfe/Input/WaterBudget/AMC5_yesterday_inner.py", line 1, in <module>
        from numba.pycc import CC
    ImportError: No module named numba.pycc

    ----------------------------------------
```

I added [`numba=0.38.1`](https://github.com/drexeleds/gwlf-e/blob/9b1f3902b935af4b7047c2a5f859c9f1844406ea/requirements.txt#L7) to my `requirements.txt` to install it, but I'm still getting that error. Any ideas?

**Comment by rajadain** [comment]
Hmm. Had to add _all_ the requirements in `requirements.txt` to get it to work. Would be nice to include those in `setup.py`. Also would be nice to have the same versions in `setup.py` as in `requirements.txt` (currently they differ for `numba` and `numpy`).

**Comment by mudkipmaster** [comment]
You're totally right, I didn't notice that. I have updated setup.py, but should I leave requirements.txt in place? It seems there's some disagreement online about the best practice.

**Comment by rajadain** [comment]
Thanks. I'll try and reduce the explicitly added dependencies one at a time and arrive on the minimum set required for integration. If there need to be further changes made here, I'll let you know.

**Comment by rajadain** [comment]
Since I can't make an issue on the https://github.com/drexeleds/gwlf-e fork, I'll just add this as a comment here.

## AttributeError: 'DataModel' object has no attribute 'NYrs'

When integrating GWLF-E in MMW in [this method](https://github.com/WikiWatershed/model-my-watershed/blob/5a896518a37fcd2314add65e5af344955c6e7dc2/src/mmw/apps/modeling/tasks.py#L426-L440), for the Lower Schuylkill HUC-10 we generate this input: [mapshed_data.json.txt](https://github.com/WikiWatershed/gwlf-e/files/2228097/mapshed_data.json.txt) (in JSON, actual input is a Python dictionary, so just `json.loads(mapshed_data.json.txt)` to get it), and for it get this error:

```
[2018-07-25 09:36:09,417: ERROR/ForkPoolWorker-1] Task f105dd82-6791-4e21-b73d-1cba7d464e41 run from job 4145 raised exception: 'DataModel' object has no attribute 'NYrs'
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 374, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 629, in __protected_call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/statsd/client.py", line 26, in wrapper
    return f(*args, **kw)
  File "/opt/app/apps/modeling/tasks.py", line 344, in run_gwlfe
    output = to_gms_file(model_input)
  File "/opt/app/apps/modeling/tasks.py", line 438, in to_gms_file
    writer.write(pre_z)
  File "/usr/local/lib/python2.7/dist-packages/gwlfe/Parser.py", line 1292, in write
    UnSatStorCarryover_f(z.NYrs, z.DaysMonth, z.Temp, z.InitSnow_0, z.Prec, z.NRur, z.NUrb, z.Area, z.CNI_0,
AttributeError: 'DataModel' object has no attribute 'NYrs'
```

Can this be fixed and added as a test please?

**Comment by drexeleds** [comment]
So the issue here is that the old code relied on variables being instantiated before the model is run. In the new code, we don't store these variables in `z` anymore, but just pass them where they are needed. When we rewrote GMSWriter, we wrote it such that it outputs the result of the model (and actually added the ability to do this via a command line argument). So where before when you ran GMSWriter before running the model in the old code there was already a zero for `UnsatStorCarryover`, we wanted to output the value after running the model which depends on variables such as `NYrs` which are created by running `Parser`. 

TL;DR:
There are three potential outputs of GMSWriter:

- GMS file before Parser is run
- GMS file after Parser is run
- GMS file after the model is run

Which of these three do you use? Currently we output the 3rd one, but that assumed that Parser was called before which is why you see an error

**Comment by rajadain** [comment]
Given that [our code](https://github.com/WikiWatershed/model-my-watershed/blob/3d4361f3273c1df507079bec33747a679e7f060d/src/mmw/apps/modeling/tasks.py#L426-L440) is this:

```python
pre_z = Parser.DataModel(mapshed_data)
output = StringIO()
writer = Parser.GmsWriter(output)
writer.write(pre_z)
```

We're running `GmsWriter` after initializing the Parser with `Parser.DataModel()`. We use this to generate a GMS file from the JSON representation before the model is run, and then [run the model](https://github.com/WikiWatershed/model-my-watershed/blob/3d4361f3273c1df507079bec33747a679e7f060d/src/mmw/apps/modeling/tasks.py#L332-L353) on that.

So we need the `GmsWriter` to work before the model has been run.

**Comment by drexeleds** [comment]
I have enabled issues on the drexeleds repository and copied our discussion, see: https://github.com/drexeleds/gwlf-e/issues/5
